scipy/ndimage/tests/test_interpolation.py

```diff
@@ -1061,151 +1207,158 @@ def test_shift_nearest_boundary(self, order, prefilter):
-                              [9, 10, 11, 12]], dtype=dtype)
+    def test_zoom_affine01(self, order, dtype, xp):
+        data = xp.asarray([[1, 2, 3, 4],
+                           [5, 6, 7, 8],
+                           [9, 10, 11, 12]], dtype=dtype)
         if data.dtype.kind == 'c':
             data -= 1j * data
         with suppress_warnings() as sup:
             sup.filter(UserWarning,
                        'The behavior of affine_transform with a 1-D array .* '
                        'has changed')
-            out = ndimage.affine_transform(data, [0.5, 0.5], 0,
+            out = ndimage.affine_transform(data, xp.asarray([0.5, 0.5]), 0,
                                            (6, 8), order=order)
         assert_array_almost_equal(out[::2, ::2], data)
 
-    def test_zoom_infinity(self):
+    def test_zoom_infinity(self, xp):
         # Ticket #1419 regression test
         dim = 8
-        ndimage.zoom(np.zeros((dim, dim)), 1. / dim, mode='nearest')
+        ndimage.zoom(xp.zeros((dim, dim)), 1. / dim, mode='nearest')
 
-    def test_zoom_zoomfactor_one(self):
+    def test_zoom_zoomfactor_one(self, xp):
         # Ticket #1122 regression test
-        arr = np.zeros((1, 5, 5))
+        arr = xp.zeros((1, 5, 5))
         zoom = (1.0, 2.0, 2.0)
 
         out = ndimage.zoom(arr, zoom, cval=7)
-        ref = np.zeros((1, 10, 10))
+        ref = xp.zeros((1, 10, 10))
         assert_array_almost_equal(out, ref)
 
-    def test_zoom_output_shape_roundoff(self):
-        arr = np.zeros((3, 11, 25))
+    def test_zoom_output_shape_roundoff(self, xp):
+        arr = xp.zeros((3, 11, 25))
         zoom = (4.0 / 3, 15.0 / 11, 29.0 / 25)
         out = ndimage.zoom(arr, zoom)
-        assert_array_equal(out.shape, (4, 15, 29))
+        assert out.shape == (4, 15, 29)
 
     @pytest.mark.parametrize('zoom', [(1, 1), (3, 5), (8, 2), (8, 8)])
     @pytest.mark.parametrize('mode', ['nearest', 'constant', 'wrap', 'reflect',
                                       'mirror', 'grid-wrap', 'grid-mirror',
                                       'grid-constant'])
-    def test_zoom_by_int_order0(self, zoom, mode):
+    def test_zoom_by_int_order0(self, zoom, mode, xp):
         # order 0 zoom should be the same as replication via np.kron
         # Note: This is not True for general x shapes when grid_mode is False,
         #       but works here for all modes because the size ratio happens to
         #       always be an integer when x.shape = (2, 2).
-        x = np.array([[0, 1],
-                         [2, 3]], dtype=float)
-        # x = np.arange(16, dtype=float).reshape(4, 4)
+        x = xp.asarray([[0, 1],
+                        [2, 3]], dtype=float)
+        # x = xp.arange(16, dtype=float).reshape(4, 4)
+        if is_cupy(xp):
+            expected = np.kron(x.get(), np.ones(zoom))
+            expected = xp.asarray(expected)
+        else:
+            expected = np.kron(x, np.ones(zoom))
```

Comment:
likewise here, can we just delay getting the `xp` version of `x` until we have calculated `expected`?